Проект ShareIt. Техническое задание. Часть 1. Реализация. Add-controllers

KanID92 opened the pull request.

KanID92 committed:
Проект ShareIt. Техническое задание. Часть 1. Реализация. Add-controllers.

kesch9 commented on src/main/java/ru/practicum/shareit/user/UserController.java:
```diff
@@ -1,12 +1,72 @@
+        log.info("==> POST. Creating new user: {}", userDto.toString());
+        validationService.validateUserDtoCreate(userDto);
+        User user = UserDtoMapper.fromDto(userDto);
+        UserDto newUserDto = UserDtoMapper.toDto(userService.create(user));
```
Обязательно
Валидация и маппинг должен быть в сервисном классе, нужно перенести туда это. Это же замечание и для остальных контроллеров

kesch9 commented on src/main/java/ru/practicum/shareit/user/UserController.java:
```diff
@@ -1,12 +1,72 @@
  */
 @RestController
 @RequestMapping(path = "/users")
+@RequiredArgsConstructor
```
Обязательно
После выполнения и подготовки PR нужно удалять TODO, их не должно быть на ревью

kesch9 commented on src/main/java/ru/practicum/shareit/user/dto/UserDtoMapper.java:
```diff
@@ -0,0 +1,30 @@
+        return user;
+    }
+
+    public static List<UserDto> toDtos(List<User> users) {
```
Обязательно
Преобразований списка в список, не нужно создавать, у вас есть метод toDto его вызывайте. Нужно использовать уже stream + map

kesch9 commented on src/main/java/ru/practicum/shareit/user/model/User.java:
```diff
@@ -0,0 +1,16 @@
+
+/**
+ *
+ */
```
Обязательно
Пустых комментариев не должно быть

kesch9 commented on src/main/java/ru/practicum/shareit/user/repository/InMemoryUserRepository.java:
```diff
@@ -0,0 +1,83 @@
+import java.util.*;
+
+
+@Slf4j()
```
Обязательно
Скобки не нужны  и  логирования в репизториях нет, можно и не добавлять эту аннотацию

kesch9 commented on src/main/java/ru/practicum/shareit/user/repository/InMemoryUserRepository.java:
```diff
@@ -0,0 +1,83 @@
+public class InMemoryUserRepository implements UserRepository {
+
+    Map<Long, User> users = new HashMap<>();
+    private static long idUserCounter = 0;
```
Обязательно
Счетчик не статичный в репозиториях

kesch9 commented on src/main/java/ru/practicum/shareit/user/repository/InMemoryUserRepository.java:
```diff
@@ -0,0 +1,83 @@
+@Repository
+public class InMemoryUserRepository implements UserRepository {
+
+    Map<Long, User> users = new HashMap<>();
```
Обязательно
Забываете про private final

kesch9 commented on src/main/java/ru/practicum/shareit/user/repository/InMemoryUserRepository.java:
```diff
@@ -0,0 +1,83 @@
+    public void delete(long userId) {
+        try {
+            users.remove(userId);
+        } catch (NullPointerException e) {
```
Обязательно
NullPointerException не кидается при удалении из мапы, данный catch не нужен

kesch9 commented on src/main/java/ru/practicum/shareit/user/repository/InMemoryUserRepository.java:
```diff
@@ -0,0 +1,83 @@
+            users.get(user.getId());
+            users.put(user.getId(), user);
+            return users.get(user.getId());
+        } catch (NullPointerException e) {
```
Обязательно
Данный catch не нужен вы же в 30 строчке проверили, что такая запись есть

kesch9 commented on src/main/java/ru/practicum/shareit/user/repository/InMemoryUserRepository.java:
```diff
@@ -0,0 +1,83 @@
+            return users.get(user.getId());
+        } catch (NullPointerException e) {
+            throw new NotFoundException("Update: not found User with id = " + user.getId());
+        } catch (ConflictException e) {
```
Обязательно
Ловить ошибку и затем ее же кидать не нужно, она и так дальше прокинеться. Данный catch лишний

kesch9 commented on src/main/java/ru/practicum/shareit/user/dto/UserDtoMapper.java:
```diff
@@ -0,0 +1,30 @@
+import java.util.ArrayList;
+import java.util.List;
+
+public class UserDtoMapper {
```
Обязательно
Для утильных классов нужно проставлять
@UtilityClass
или делать private конструктор. Также мапперы должны лежать в пакете mapper(или другое название), но не в dto

kesch9 commented on src/main/java/ru/practicum/shareit/item/repository/InMemoryItemRepository.java:
```diff
@@ -0,0 +1,82 @@
+        if (query.isBlank()) {
+            return List.of();
+        }
+        for (Item item : items.values()) {
```
Обязательно
Используем stream+filter

kesch9 commented on src/main/java/ru/practicum/shareit/item/repository/InMemoryItemRepository.java:
```diff
@@ -0,0 +1,82 @@
+    public List<Item> getAllUserItems(long userId) {
+        List<Item> userItems = new ArrayList<>();
+        for (Item item : items.values()) {
+            if (item.getOwnerId() == userId) {
```
Обязательно
Используем stream+filter

kesch9 commented on src/main/java/ru/practicum/shareit/item/repository/InMemoryItemRepository.java:
```diff
@@ -0,0 +1,82 @@
+public class InMemoryItemRepository implements ItemRepository {
+
+    private final Map<Long, Item> items = new HashMap<>();
+    private static long idItemCounter = 0;
```
Обязательно
Счетчик id не статичный

kesch9 commented on src/main/java/ru/practicum/shareit/item/service/ItemServiceImpl.java:
```diff
@@ -0,0 +1,66 @@
+    private final ItemRepository itemRepository;
+    private final UserService userService;
+
+    public Item add(Item item) {
```
Обязательно
Не над всеми методами проставлено @Override

kesch9 commented on src/main/java/ru/practicum/shareit/item/service/ItemServiceImpl.java:
```diff
@@ -0,0 +1,66 @@
+        }
+        try {
+            return itemRepository.update(item);
+        } catch (NotFoundException e) {
```
Обязательно
Ловить NotFoundException и затем кидать заново его не надо

kesch9 commented on src/main/java/ru/practicum/shareit/item/service/ItemServiceImpl.java:
```diff
@@ -0,0 +1,66 @@
+        userService.getById(item.getOwnerId());
+        try {
+            return itemRepository.add(item);
+        } catch (NotFoundException e) {
```
Обязательно
Ловить NotFoundException и затем кидать заново его не надо

kesch9 commented on src/main/java/ru/practicum/shareit/validation/ValidationServiceImpl.java:
```diff
@@ -0,0 +1,57 @@
+import java.util.Optional;
+
+@Service
+public class ValidationServiceImpl implements ValidationService {
```
Можно лучше
Валидацию уже надо делать на аннотациях, вместо отдельных сервисах.

kesch9 reviewed (commented):
Нужно доработать код

KanID92 committed:
Проект ShareIt. Техническое задание. Часть 1. Реализация. Add-controllers.
Коорректировка по ревью:
- валидация и маппинг перенести в сервисный класс;
- удалить TODO;
- заменить циклы на Stream;
- убрать излишний метод List<USerDto> toDtos(List<User> users);
- убрать статичность счетчиков id;
- дополнить private, final в поле хранения users в UserRepository;
- убрать излишние catch ошибок;
- обозначить мапперы как @UtilityClass;
- дополнить пропущенные @Override;
- перенести мапперы в отдельный пакет (убрать из пакетов dto);
- выполнить рефакторинг валидации: выполнение через аннотации;

KanID92 committed:
Проект ShareIt. Техническое задание. Часть 1. Реализация. Add-controllers.
Коорректировка по ревью:
- валидация и маппинг перенести в сервисный класс;
- удалить TODO;
- заменить циклы на Stream;
- убрать излишний метод List<USerDto> toDtos(List<User> users);
- убрать статичность счетчиков id;
- дополнить private, final в поле хранения users в UserRepository;
- убрать излишние catch ошибок;
- обозначить мапперы как @UtilityClass;
- дополнить пропущенные @Override;
- перенести мапперы в отдельный пакет (убрать из пакетов dto);
- выполнить рефакторинг валидации: выполнение через аннотации;

KanID92 committed:
Merge remote-tracking branch 'origin/add-controllers' into add-controllers

kesch9 commented on src/main/java/ru/practicum/shareit/item/model/Item.java:
```diff
@@ -1,7 +1,32 @@
 public class Item {
+
+    @Null(groups = Marker.OnCreate.class, message = "Created Item can't already have id")
+    @NotNull(groups = Marker.OnUpdate.class, message = "Update Item must have id")
```
Обязательно
Валидировать модели не нужно, они будут валидироваться на уровне БД, уберите валидационные аннотации с моделей к следующему спринту

kesch9 reviewed (approved):
Approve

KanID92 committed:
- убрать аннотации валидации с моделей;
